{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO\n",
    "\n",
    "last time bois\n",
    "* 1) just add a few keywords ( like they did )\n",
    "* 2\n",
    "\n",
    "* print pyserini query builder - can we replicate this?\n",
    "* compare tfidf, previously best version, full-text? tfidf3\n",
    "* create final.fulltext.qq\n",
    "* create final.fulltext.qt\n",
    "* create final.fusion.qt\n",
    "\n",
    "\n",
    "ok new plans.\n",
    "* compare round3 round4 results\n",
    "* do better task weighting to use all tfidf terms\n",
    "* prepare round5 submission\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "* test with top 3 tfidf\n",
    "* test na prep final results\n",
    "\n",
    "\n",
    "* Code to generate our topics for Anserini (extend UDel, and alternative to UDel)\n",
    "* Test task_short_title\n",
    "* Test task_NER_long\n",
    "* Make sure we replicate https://github.com/castorini/anserini/blob/master/src/main/python/trec-covid/generate_query_udel.py\n",
    "* Celebrate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndentationError",
     "evalue": "unexpected indent (<ipython-input-1-34b4b3e89af8>, line 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"<ipython-input-1-34b4b3e89af8>\"\u001b[0;36m, line \u001b[0;32m3\u001b[0m\n\u001b[0;31m    topics nDCG@10   J@10 nDCG@20   J@20     AP   R@1k   J@1k MD5\u001b[0m\n\u001b[0m    ^\u001b[0m\n\u001b[0;31mIndentationError\u001b[0m\u001b[0;31m:\u001b[0m unexpected indent\n"
     ]
    }
   ],
   "source": [
    "# Evaluation results w/ src/main/resources/topics-and-qrels/qrels.covid-round4-cumulative.txt\n",
    "\n",
    "                                            topics nDCG@10   J@10 nDCG@20   J@20     AP   R@1k   J@1k MD5\n",
    "b'map                   \\tall\\t0.1876\\nrecall_1000           \\tall\\t0.4525\\nndcg_cut_10           \\tall\\t0.4580\\n'\n",
    "anserini.covid-r5.abstract.qq.bm25.txt          50  0.4580 0.5880  0.4379 0.5940 0.1876 0.4525 0.2264 5ad088eabaf936384c8962f92348157e\n",
    "b'map                   \\tall\\t0.2018\\nrecall_1000           \\tall\\t0.4714\\nndcg_cut_10           \\tall\\t0.4912\\n'\n",
    "anserini.covid-r5.abstract.qdel.bm25.txt        50  0.4912 0.6240  0.4596 0.6040 0.2018 0.4714 0.2351 6e26623d77fdec1010b7006565a8dd20\n",
    "b'map                   \\tall\\t0.1286\\nrecall_1000           \\tall\\t0.3758\\nndcg_cut_10           \\tall\\t0.3240\\n'\n",
    "anserini.covid-r5.full-text.qq.bm25.txt         50  0.3240 0.5660  0.3055 0.5250 0.1286 0.3758 0.2171 8c311cec6f9b89f77a60882c06b7e92f\n",
    "b'map                   \\tall\\t0.1754\\nrecall_1000           \\tall\\t0.4368\\nndcg_cut_10           \\tall\\t0.4634\\n'\n",
    "anserini.covid-r5.full-text.qdel.bm25.txt       50  0.4634 0.6460  0.4387 0.6280 0.1754 0.4368 0.2425 b2f8026578feb916d7f29e407021db98\n",
    "b'map                   \\tall\\t0.1680\\nrecall_1000           \\tall\\t0.4790\\nndcg_cut_10           \\tall\\t0.4077\\n'\n",
    "anserini.covid-r5.paragraph.qq.bm25.txt         50  0.4077 0.6160  0.3907 0.5920 0.1680 0.4790 0.2721 4701bc67333302abc1045cea59c5978c\n",
    "b'map                   \\tall\\t0.1845\\nrecall_1000           \\tall\\t0.4978\\nndcg_cut_10           \\tall\\t0.4918\\n'\n",
    "anserini.covid-r5.paragraph.qdel.bm25.txt       50  0.4918 0.6440  0.4569 0.6250 0.1845 0.4978 0.2779 3e1c3ce9ca42ed633b6db681c61c8439\n",
    "b'map                   \\tall\\t0.2041\\nrecall_1000           \\tall\\t0.5024\\nndcg_cut_10           \\tall\\t0.4696\\n'\n",
    "anserini.covid-r5.fusion1.txt                   50  0.4696 0.6520  0.4539 0.6490 0.2041 0.5024 0.2747 889cab8bee51759554ba974c6008d1ab\n",
    "b'map                   \\tall\\t0.2300\\nrecall_1000           \\tall\\t0.5371\\nndcg_cut_10           \\tall\\t0.5077\\n'\n",
    "anserini.covid-r5.fusion2.txt                   50  0.5077 0.6800  0.4956 0.6690 0.2300 0.5371 0.2850 2031fd0aeec7831da79e5568cb86e2a1\n",
    "b'map                   \\tall\\t0.2000\\nrecall_1000           \\tall\\t0.4721\\nndcg_cut_10           \\tall\\t0.4543\\n'\n",
    "anserini.covid-r5.abstract.ruir1.txt            50  0.4543 0.5800  0.4510 0.5980 0.2000 0.4721 0.2331 5d7a3844ac9d88778649a5faa0f4b78e\n",
    "b'map                   \\tall\\t0.1996\\nrecall_1000           \\tall\\t0.4719\\nndcg_cut_10           \\tall\\t0.4688\\n'\n",
    "anserini.covid-r5.abstract.ruir2.txt            50  0.4688 0.6080  0.4511 0.5840 0.1996 0.4719 0.2335 754f1bfa355ca4fda80cd27d4e4011e8\n",
    "b'map                   \\tall\\t0.1473\\nrecall_1000           \\tall\\t0.3948\\nndcg_cut_10           \\tall\\t0.3621\\n'\n",
    "anserini.covid-r5.abstract.ruir3.txt            50  0.3621 0.5080  0.3447 0.4810 0.1473 0.3948 0.1902 10fe8eeacf7b8924e9142130708a1fcd\n",
    "b'map                   \\tall\\t0.0066\\nrecall_1000           \\tall\\t0.0831\\nndcg_cut_10           \\tall\\t0.0126\\n'\n",
    "anserini.covid-r5.paragraph.ruir1.txt           50  0.0126 0.0140  0.0166 0.0210 0.0066 0.0831 0.0272 797d44e96553a5583f18e8a10bc73b07\n",
    "b'map                   \\tall\\t0.0067\\nrecall_1000           \\tall\\t0.0829\\nndcg_cut_10           \\tall\\t0.0151\\n'\n",
    "anserini.covid-r5.paragraph.ruir2.txt           50  0.0151 0.0180  0.0151 0.0180 0.0067 0.0829 0.0273 7a3f64efd2b62b3dd8819999efb60a93\n",
    "b'map                   \\tall\\t0.0052\\nrecall_1000           \\tall\\t0.0717\\nndcg_cut_10           \\tall\\t0.0115\\n'\n",
    "anserini.covid-r5.paragraph.ruir3.txt           50  0.0115 0.0200  0.0197 0.0310 0.0052 0.0717 0.0218 e4cd0298759bbf070a2905f4dc7d37e1\n",
    "b'map                   \\tall\\t0.1606\\nrecall_1000           \\tall\\t0.4213\\nndcg_cut_10           \\tall\\t0.4413\\n'\n",
    "anserini.covid-r5.full-text.ruir1.txt           50  0.4413 0.6040  0.4123 0.5860 0.1606 0.4213 0.2368 00eb140369298869f95d7fc36148b22b\n",
    "b'map                   \\tall\\t0.1641\\nrecall_1000           \\tall\\t0.4252\\nndcg_cut_10           \\tall\\t0.4406\\n'\n",
    "anserini.covid-r5.full-text.ruir2.txt           50  0.4406 0.6160  0.4168 0.5930 0.1641 0.4252 0.2377 1fb0854afeefa96d64062a230155dacd\n",
    "b'map                   \\tall\\t0.1300\\nrecall_1000           \\tall\\t0.3694\\nndcg_cut_10           \\tall\\t0.3537\\n'\n",
    "anserini.covid-r5.full-text.ruir3.txt           50  0.3537 0.5620  0.3452 0.5430 0.1300 0.3694 0.2080 6cc1bbc8c1ee49939cf0105b901053dc\n",
    "b'map                   \\tall\\t0.1729\\nrecall_1000           \\tall\\t0.4555\\nndcg_cut_10           \\tall\\t0.4959\\n'\n",
    "ruir.fusion1.txt                                50  0.4959 0.6440  0.4795 0.6330 0.1729 0.4555 0.2228 0d4bf7301f232adc7f457a1d03a0672f\n",
    "b'map                   \\tall\\t0.1738\\nrecall_1000           \\tall\\t0.4545\\nndcg_cut_10           \\tall\\t0.5020\\n'\n",
    "ruir.fusion2.txt                                50  0.5020 0.6460  0.4876 0.6390 0.1738 0.4545 0.2232 9a3b67b25e321630b3ea4a20b6d17447\n",
    "b'map                   \\tall\\t0.1322\\nrecall_1000           \\tall\\t0.3811\\nndcg_cut_10           \\tall\\t0.3916\\n'\n",
    "ruir.fusion3.txt                                50  0.3916 0.5580  0.3872 0.5460 0.1322 0.3811 0.1859 8635ef111497eac95cef1c9f9fb06242\n",
    "b'map                   \\tall\\t0.2629\\nrecall_1000           \\tall\\t0.5505\\nndcg_cut_10           \\tall\\t0.6177\\n'\n",
    "anserini.covid-r5.abstract.qdel.bm25+rm3Rf.txt  50  0.6177 0.6620  0.5738 0.6510 0.2629 0.5505 0.2562 86f703e630e28bf2bedc2e4ebefb3461\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: sklearn in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (0.0)\n",
      "Requirement already satisfied: pyserini in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (0.9.4.0)\n",
      "Requirement already satisfied: scispacy in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (0.3.0)\n",
      "Requirement already satisfied: spacy in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (2.3.2)\n",
      "Requirement already satisfied: nltk in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (3.4.5)\n",
      "Requirement already satisfied: pandas in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (0.25.1)\n",
      "Requirement already satisfied: numpy in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (1.16.5)\n",
      "Requirement already satisfied: scipy in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (1.3.1)\n",
      "Collecting rank_bm25\n",
      "  Using cached https://files.pythonhosted.org/packages/16/5a/23ed3132063a0684ea66fb410260c71c4ffda3b99f8f1c021d1e245401b5/rank_bm25-0.2.1-py3-none-any.whl\n",
      "Requirement already satisfied: scikit-learn in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from sklearn) (0.21.3)\n",
      "Requirement already satisfied: Cython in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from pyserini) (0.29.13)\n",
      "Requirement already satisfied: pyjnius in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from pyserini) (1.3.0)\n",
      "Requirement already satisfied: pysbd in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from scispacy) (0.3.3)\n",
      "Requirement already satisfied: requests<3.0.0conllu,>=2.0.0 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from scispacy) (2.22.0)\n",
      "Requirement already satisfied: joblib in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from scispacy) (0.13.2)\n",
      "Requirement already satisfied: nmslib>=1.7.3.6 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from scispacy) (2.0.6)\n",
      "Requirement already satisfied: tqdm<5.0.0,>=4.38.0 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from spacy) (4.50.2)\n",
      "Requirement already satisfied: preshed<3.1.0,>=3.0.2 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from spacy) (3.0.2)\n",
      "Requirement already satisfied: cymem<2.1.0,>=2.0.2 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from spacy) (2.0.3)\n",
      "Requirement already satisfied: blis<0.5.0,>=0.4.0 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from spacy) (0.4.1)\n",
      "Requirement already satisfied: thinc==7.4.1 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from spacy) (7.4.1)\n",
      "Requirement already satisfied: wasabi<1.1.0,>=0.4.0 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from spacy) (0.8.0)\n",
      "Requirement already satisfied: setuptools in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from spacy) (41.4.0)\n",
      "Requirement already satisfied: plac<1.2.0,>=0.9.6 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from spacy) (1.1.3)\n",
      "Requirement already satisfied: catalogue<1.1.0,>=0.0.7 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from spacy) (1.0.0)\n",
      "Requirement already satisfied: srsly<1.1.0,>=1.0.2 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from spacy) (1.0.2)\n",
      "Requirement already satisfied: murmurhash<1.1.0,>=0.28.0 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from spacy) (1.0.2)\n",
      "Requirement already satisfied: six in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from nltk) (1.12.0)\n",
      "Requirement already satisfied: pytz>=2017.2 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from pandas) (2019.3)\n",
      "Requirement already satisfied: python-dateutil>=2.6.1 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from pandas) (2.8.0)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from requests<3.0.0conllu,>=2.0.0->scispacy) (2019.9.11)\n",
      "Requirement already satisfied: chardet<3.1.0,>=3.0.2 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from requests<3.0.0conllu,>=2.0.0->scispacy) (3.0.4)\n",
      "Requirement already satisfied: urllib3!=1.25.0,!=1.25.1,<1.26,>=1.21.1 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from requests<3.0.0conllu,>=2.0.0->scispacy) (1.24.2)\n",
      "Requirement already satisfied: idna<2.9,>=2.5 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from requests<3.0.0conllu,>=2.0.0->scispacy) (2.8)\n",
      "Requirement already satisfied: psutil in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from nmslib>=1.7.3.6->scispacy) (5.6.3)\n",
      "Requirement already satisfied: pybind11>=2.2.3 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from nmslib>=1.7.3.6->scispacy) (2.5.0)\n",
      "Requirement already satisfied: importlib-metadata>=0.20; python_version < \"3.8\" in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from catalogue<1.1.0,>=0.0.7->spacy) (0.23)\n",
      "Requirement already satisfied: zipp>=0.5 in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from importlib-metadata>=0.20; python_version < \"3.8\"->catalogue<1.1.0,>=0.0.7->spacy) (0.6.0)\n",
      "Requirement already satisfied: more-itertools in c:\\users\\allemaal\\anaconda3\\lib\\site-packages (from zipp>=0.5->importlib-metadata>=0.20; python_version < \"3.8\"->catalogue<1.1.0,>=0.0.7->spacy) (7.2.0)\n",
      "Installing collected packages: rank-bm25\n",
      "Successfully installed rank-bm25-0.2.1\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "!{sys.executable} -m pip install sklearn pyserini scispacy spacy nltk pandas numpy scipy rank_bm25"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " 1 rflu6axg        4.81110\n",
      " 2 6f3up1h0        4.73690\n",
      " 3 gq3965se        4.70890\n",
      " 4 u13oge9m        4.64420\n",
      " 5 z6f56n7h        4.55250\n",
      " 6 ipk43u69        4.55020\n",
      " 7 r6wwpfcy        4.45910\n",
      " 8 js9ek5me        4.43840\n",
      " 9 d45xh7l0        4.43830\n",
      "10 r68l7g2e        4.41370\n",
      "\n",
      " 1 rflu6axg        7.88680\n",
      " 2 u13oge9m        7.76570\n",
      " 3 gq3965se        7.75320\n",
      " 4 6f3up1h0        7.57950\n",
      " 5 r6wwpfcy        7.32980\n",
      " 6 thg4fqi2        7.31170\n",
      " 7 r68l7g2e        7.27710\n",
      " 8 4ywt0yqn        7.25290\n",
      " 9 rpf02tc0        7.21640\n",
      "10 yrz8s11w        7.20100\n"
     ]
    }
   ],
   "source": [
    "#quick query term test\n",
    "from pyserini.search import SimpleSearcher\n",
    "#from pyserini.index import pyutils\n",
    "#from pyserini.analysis import pyanalysis\n",
    "#from pyserini.search import pysearch\n",
    "from pyserini.search import querybuilder\n",
    "#from pyserini.pyclass import autoclass\n",
    "#from pyserini.analysis.pyanalysis import get_lucene_analyzer\n",
    "\n",
    "prefix = 'C:/Users/Allemaal/Desktop/ubuntu/Desktop/anserini/indexes/'\n",
    "\n",
    "index_loc = prefix + 'lucene-index-cord19-full-text-2020-07-16'\n",
    "#'/media/tmschoegje/Data1/indexes/lucene-index-cord19-full-text-2020-05-19/'\n",
    "searcher = SimpleSearcher(index_loc)\n",
    "#index_utils = pyutils.IndexReaderUtils(index_loc)\n",
    "\n",
    "\n",
    "hits = searcher.search('corona origin')\n",
    "\n",
    "# Print the first 10 hits:\n",
    "for i in range(0, 10):\n",
    "    print(f'{i+1:2} {hits[i].docid:15} {hits[i].score:.5f}')\n",
    "    \n",
    "    \n",
    "hits = searcher.search('corona origin origin origin')\n",
    "print()\n",
    "for i in range(0, 10):\n",
    "    print(f'{i+1:2} {hits[i].docid:15} {hits[i].score:.5f}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "org.apache.lucene.search.BooleanQuery$Builder@542e560f\n",
      "contents:corona contents:corona contents:origin\n",
      " 1 rflu6axg        8.08440\n",
      " 2 6f3up1h0        8.05250\n",
      " 3 z6f56n7h        7.90320\n",
      " 4 gq3965se        7.89550\n",
      " 5 ipk43u69        7.83340\n",
      " 6 gzxu0r08        7.82560\n",
      " 7 woamqlyg        7.79640\n",
      " 8 u13oge9m        7.72760\n",
      " 9 js9ek5me        7.71020\n",
      "10 zbsidnsc        7.70580\n",
      "\n",
      " 1 rflu6axg        7.88680\n",
      " 2 u13oge9m        7.76570\n",
      " 3 gq3965se        7.75320\n",
      " 4 6f3up1h0        7.57950\n",
      " 5 r6wwpfcy        7.32980\n",
      " 6 thg4fqi2        7.31170\n",
      " 7 r68l7g2e        7.27710\n",
      " 8 4ywt0yqn        7.25290\n",
      " 9 rpf02tc0        7.21640\n",
      "10 yrz8s11w        7.20100\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n 1 rflu6axg        2.68350\\n 2 6f3up1h0        2.67890\\n 3 z6f56n7h        2.64580\\n 4 gzxu0r08        2.62060\\n 5 ipk43u69        2.60450\\n 6 js9ek5me        2.58240\\n 7 u13oge9m        2.57850\\n 8 zbsidnsc        2.57740\\n 9 feq9grdc        2.54000\\n10 yh7mxwkg        2.53230\\n\\n'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#TEST what does querybuilder look like?\n",
    "from pyserini.search import SimpleSearcher\n",
    "from pyserini.search import querybuilder\n",
    "\n",
    "index_loc = prefix + 'lucene-index-cord19-full-text-2020-07-16/'\n",
    "searcher = SimpleSearcher(index_loc)\n",
    "\n",
    "builder = querybuilder.get_boolean_query_builder()\n",
    "\n",
    "terms = ['corona', 'corona', 'origin']\n",
    "bm25_split = 1\n",
    "for b in terms:\n",
    "    b = querybuilder.get_term_query(b)\n",
    "    boostquery = querybuilder.get_boost_query(b, bm25_split/len(terms))\n",
    "    builder.add(b, querybuilder.JBooleanClauseOccur['should'].value)\n",
    "\n",
    "print(builder.toString())\n",
    "query = builder.build()\n",
    "print(query.toString())\n",
    "\n",
    "hits = searcher.search(query, 50)\n",
    "#hits = searcher.search('corona origin')\n",
    "\n",
    "# Print the first 10 hits:\n",
    "for i in range(0, 10):\n",
    "    print(f'{i+1:2} {hits[i].docid:15} {hits[i].score:.5f}')\n",
    "    \n",
    "    \n",
    "hits = searcher.search('corona origin origin origin')\n",
    "print()\n",
    "for i in range(0, 10):\n",
    "    print(f'{i+1:2} {hits[i].docid:15} {hits[i].score:.5f}')\n",
    "\"\"\"\n",
    " 1 rflu6axg        2.68350\n",
    " 2 6f3up1h0        2.67890\n",
    " 3 z6f56n7h        2.64580\n",
    " 4 gzxu0r08        2.62060\n",
    " 5 ipk43u69        2.60450\n",
    " 6 js9ek5me        2.58240\n",
    " 7 u13oge9m        2.57850\n",
    " 8 zbsidnsc        2.57740\n",
    " 9 feq9grdc        2.54000\n",
    "10 yh7mxwkg        2.53230\n",
    "\n",
    "\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scispacy\n",
    "import spacy\n",
    "import xml.etree.ElementTree as Et\n",
    "import nltk\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "task_1_short = \"transmission incubation environment\"\n",
    "task_2_short = \"risk factors\"\n",
    "task_3_short = \"genetics origin evolution\"\n",
    "task_4_short = \"vaccines therapeutics\"\n",
    "task_5_short = \"medical care\"\n",
    "task_6_short = \"effectiveness non pharmaceutical interventions\"\n",
    "task_7_short = \"diagnostics surveillance\"\n",
    "task_8_short = \"geographic spread\"\n",
    "task_9_short = \"ethical social considerations\"\n",
    "task_10_short = \"information sharing collaboration\"\n",
    "\n",
    "\n",
    "\n",
    "task_1 = \"What is known about transmission, incubation, and environmental stability? What do we know about natural history, transmission, and diagnostics for the virus? What have we learned about infection prevention and control? Range of incubation periods for the disease in humans (and how this varies across age and health status) and how long individuals are contagious, even after recovery. Prevalence of asymptomatic shedding and transmission (e.g., particularly children). Seasonality of transmission. Physical science of the coronavirus (e.g., charge distribution, adhesion to hydrophilic/phobic surfaces, environmental survival to inform decontamination efforts for affected areas and provide information about viral shedding). Persistence and stability on a multitude of substrates and sources (e.g., nasal discharge, sputum, urine, fecal matter, blood). Persistence of virus on surfaces of different materials (e,g., copper, stainless steel, plastic). Natural history of the virus and shedding of it from an infected person. Implementation of diagnostics and products to improve clinical processes. Disease models, including animal models for infection, disease and transmission. Tools and studies to monitor phenotypic change and potential adaptation of the virus. Immune response and immunity. Effectiveness of movement control strategies to prevent secondary transmission in health care and community settings. Effectiveness of personal protective equipment (PPE) and its usefulness to reduce risk of transmission in health care and community settings. Role of the environment in transmission\"\n",
    "task_2 = \"What do we know about COVID-19 risk factors? What have we learned from epidemiological studies? Data on potential risks factors. Smoking, pre-existing pulmonary disease. Co-infections (determine whether co-existing respiratory/viral infections make the virus more transmissible or virulent) and other co-morbidities. Neonates and pregnant women. Socio-economic and behavioral factors to understand the economic impact of the virus and whether there were differences. Transmission dynamics of the virus, including the basic reproductive number, incubation period, serial interval, modes of transmission and environmental factors. Severity of disease, including risk of fatality among symptomatic hospitalized patients, and high-risk patient groups. Susceptibility of populations. Public health mitigation measures that could be effective for control.\"\n",
    "task_3 = \"What do we know about virus genetics, origin, and evolution? What do we know about the virus origin and management measures at the human-animal interface? Real-time tracking of whole genomes and a mechanism for coordinating the rapid dissemination of that information to inform the development of diagnostics and therapeutics and to track variations of the virus over time. Access to geographic and temporal diverse sample sets to understand geographic distribution and genomic differences, and determine whether there is more than one strain in circulation. Multi-lateral agreements such as the Nagoya Protocol could be leveraged. Evidence that livestock could be infected (e.g., field surveillance, genetic sequencing, receptor binding) and serve as a reservoir after the epidemic appears to be over. Evidence of whether farmers are infected, and whether farmers could have played a role in the origin. Surveillance of mixed wildlife- livestock farms for SARS-CoV-2 and other coronaviruses in Southeast Asia. Experimental infections to test host range for this pathogen. Animal host(s) and any evidence of continued spill-over to humans. Socioeconomic and behavioral risk factors for this spill-over. Sustainable risk reduction strategies\"\n",
    "task_4 = \"What do we know about vaccines and therapeutics? What has been published concerning research and development and evaluation efforts of vaccines and therapeutics? Effectiveness of drugs being developed and tried to treat COVID-19 patients. Clinical and bench trials to investigate less common viral inhibitors against COVID-19 such as naproxen, clarithromycin, and minocycline that that may exert effects on viral replication. Methods evaluating potential complication of Antibody-Dependent Enhancement (ADE) in vaccine recipients. Exploration of use of best animal models and their predictive value for a human vaccine. Capabilities to discover a therapeutic (not vaccine) for the disease, and clinical effectiveness studies to discover therapeutics, to include antiviral agents. Alternative models to aid decision makers in determining how to prioritize and distribute scarce, newly proven therapeutics as production ramps up. This could include identifying approaches for expanding production capacity to ensure equitable and timely distribution to populations in need. Efforts targeted at a universal coronavirus vaccine. Efforts to develop animal models and standardize challenge studies. Efforts to develop prophylaxis clinical studies and prioritize in healthcare workers. Approaches to evaluate risk for enhanced disease after vaccination. Assays to evaluate vaccine immune response and process development for vaccines, alongside suitable animal models in conjunction with therapeutics.\"\n",
    "task_5 = \"What has been published about medical care? What has been published concerning surge capacity and nursing homes? What has been published concerning efforts to inform allocation of scarce resources? What do we know about personal protective equipment? What has been published concerning alternative methods to advise on disease management? What has been published concerning processes of care? What do we know about the clinical characterization and management of the virus? Resources to support skilled nursing facilities and long term care facilities. Mobilization of surge medical staff to address shortages in overwhelmed communities. Age-adjusted mortality data for Acute Respiratory Distress Syndrome (ARDS) with/without other organ failure – particularly for viral etiologies. Extracorporeal membrane oxygenation (ECMO) outcomes data of COVID-19 patients. Outcomes data for COVID-19 after mechanical ventilation adjusted for age. Application of regulatory standards (e.g., EUA, CLIA) and ability to adapt care to crisis standards of care level. Approaches for encouraging and facilitating the production of elastomeric respirators, which can save thousands of N95 masks. Best telemedicine practices, barriers and facilitators, and specific actions to remove/expand them within and across state boundaries. Guidance on the simple things people can do at home to take care of sick people and manage disease. Oral medications that might potentially work. Use of AI in real-time health care delivery to evaluate interventions, risk factors, and outcomes in a way that could not be done manually. Best practices and critical challenges and innovative solutions and technologies in hospital flow and organization, workforce protection, workforce allocation, community-based support resources, payment, and supply chain management to enhance capacity, efficiency, and outcomes. Efforts to define the natural history of disease to inform clinical care, public health interventions, infection prevention control, transmission, and clinical trials. Efforts to develop a core clinical outcome set to maximize usability of data across a range of trials. Efforts to determine adjunctive and supportive interventions that can improve the clinical outcomes of infected patients (e.g. steroids, high flow oxygen). \"\n",
    "task_6 = \"What do we know about the effectiveness of non-pharmaceutical interventions? What is known about equity and barriers to compliance for non-pharmaceutical interventions? Guidance on ways to scale up NPIs in a more coordinated way (e.g., establish funding, infrastructure and authorities to support real time, authoritative (qualified participants) collaboration with all states to gain consensus on consistent guidance and to mobilize resources to geographic areas where critical shortfalls are identified) to give us time to enhance our health care delivery system capacity to respond to an increase in cases. Rapid design and execution of experiments to examine and compare NPIs currently being implemented. DHS Centers for Excellence could potentially be leveraged to conduct these experiments. Rapid assessment of the likely efficacy of school closures, travel bans, bans on mass gatherings of various sizes, and other social distancing approaches. Methods to control the spread in communities, barriers to compliance and how these vary among different populations. Models of potential interventions to predict costs and benefits that take account of such factors as race, income, disability, age, geographic location, immigration status, housing status, employment status, and health insurance status. Policy changes necessary to enable the compliance of individuals with limited resources and the underserved with NPIs. Research on why people fail to comply with public health advice, even if they want to do so (e.g., social or financial costs may be too high). Research on the economic impact of this or any pandemic. This would include identifying policy and programmatic alternatives that lessen/mitigate risks to critical government services, food distribution and supplies, access to critical household supplies, and access to health diagnoses, treatment, and needed care, regardless of ability to pay.\"\n",
    "task_7 = \"What do we know about diagnostics and surveillance? What has been published concerning systematic, holistic approach to diagnostics (from the public health surveillance perspective to being able to predict clinical outcomes)? How widespread current exposure is to be able to make immediate policy recommendations on mitigation measures. Denominators for testing and a mechanism for rapidly sharing that information, including demographics, to the extent possible. Sampling methods to determine asymptomatic disease (e.g., use of serosurveys (such as convalescent samples) and early detection of disease (e.g., use of screening of neutralizing antibodies such as ELISAs). Efforts to increase capacity on existing diagnostic platforms and tap into existing surveillance platforms. Recruitment, support, and coordination of local expertise and capacity (public, private—commercial, and non-profit, including academic), including legal, ethical, communications, and operational issues. National guidance and guidelines about best practices to states (e.g., how states might leverage universities and private laboratories for testing purposes, communications to public health officials and the public). Development of a point-of-care test (like a rapid influenza test) and rapid bed-side tests, recognizing the tradeoffs between speed, accessibility, and accuracy. Rapid design and execution of targeted surveillance experiments calling for all potential testers using PCR in a defined area to start testing and report to a specific entity. These experiments could aid in collecting longitudinal samples, which are critical to understanding the impact of ad hoc local interventions (which also need to be recorded). Separation of assay development issues from instruments, and the role of the private sector to help quickly migrate assays onto those devices. Efforts to track the evolution of the virus (i.e., genetic drift or mutations) and avoid locking into specific reagents and surveillance/detection schemes. Latency issues and when there is sufficient viral load to detect the pathogen, and understanding of what is needed in terms of biological and environmental sampling. Use of diagnostics such as host response markers (e.g., cytokines) to detect early disease or predict severe disease progression, which would be important to understanding best clinical practice and efficacy of therapeutic interventions. Policies and protocols for screening and testing. Policies to mitigate the effects on supplies associated with mass testing, including swabs and reagents. Technology roadmap for diagnostics. Barriers to developing and scaling up new diagnostic tests (e.g., market forces), how future coalition and accelerator models (e.g., Coalition for Epidemic Preparedness Innovations) could provide critical funding for diagnostics, and opportunities for a streamlined regulatory environment. New platforms and technology (e.g., CRISPR) to improve response times and employ more holistic approaches to COVID-19 and future diseases. Coupling genomics and diagnostic testing on a large scale. Enhance capabilities for rapid sequencing and bioinformatics to target regions of the genome that will allow specificity for a particular variant. Enhance capacity (people, technology, data) for sequencing with advanced analytics for unknown pathogens, and explore capabilities for distinguishing naturally-occurring pathogens from intentional. One Health surveillance of humans and potential sources of future spillover or ongoing exposure for this organism and future pathogens, including both evolutionary hosts (e.g., bats) and transmission hosts (e.g., heavily trafficked and farmed wildlife and domestic food and companion species), inclusive of environmental, demographic, and occupational risk factors.\"\n",
    "task_8 = \"At the time of writing, COVID-19 has spread to at least 114 countries. With viral flu, there are often geographic variations in how the disease will spread and if there are different variations of the virus in different areas. We’d like to explore what the literature and data say about this through this Task. Are there geographic variations in the rate of COVID-19 spread? Are there geographic variations in the mortality rate of COVID-19? Is there any evidence to suggest geographic based virus mutations?\"\n",
    "task_9 = \"What has been published concerning ethical considerations for research? What has been published concerning social sciences at the outbreak response? Efforts to articulate and translate existing ethical principles and standards to salient issues in COVID-2019. Efforts to embed ethics across all thematic areas, engage with novel ethical issues that arise and coordinate to minimize duplication of oversight. Efforts to support sustained education, access, and capacity building in the area of ethics. Efforts to establish a team at WHO that will be integrated within multidisciplinary research and operational platforms and that will connect with existing and expanded global networks of social sciences. Efforts to develop qualitative assessment frameworks to systematically collect information related to local barriers and enablers for the uptake and adherence to public health measures for prevention and control. This includes the rapid identification of the secondary impacts of these measures. (e.g. use of surgical masks, modification of health seeking behaviors for SRH, school closures). Efforts to identify how the burden of responding to the outbreak and implementing public health measures affects the physical and psychological health of those providing care for Covid-19 patients and identify the immediate needs that must be addressed. Efforts to identify the underlying drivers of fear, anxiety and stigma that fuel misinformation and rumor, particularly through social media.\"\n",
    "task_10 = \"What has been published about information sharing and inter-sectoral collaboration? What has been published about data standards and nomenclature? What has been published about governmental public health? What do we know about risk communication? What has been published about communicating with high-risk populations? What has been published to clarify community measures? What has been published about equity considerations and problems of inequity? Methods for coordinating data-gathering with standardized nomenclature. Sharing response information among planners, providers, and others. Understanding and mitigating barriers to information-sharing. How to recruit, support, and coordinate local (non-Federal) expertise and capacity relevant to public health emergency response (public, private, commercial and non-profit, including academic). Integration of federal/state/local public health surveillance systems. Value of investments in baseline public health response infrastructure preparedness. Modes of communicating with target high-risk populations (elderly, health care workers). Risk communication and guidelines that are easy to understand and follow (include targeting at risk populations’ families too). Communication that indicates potential risk of disease to all population groups. Misunderstanding around containment and mitigation. Action plan to mitigate gaps and problems of inequity in the Nation’s public health capability, capacity, and funding to ensure all citizens in need are supported and can access information, surveillance, and treatment. Measures to reach marginalized and disadvantaged populations. Data systems and research priorities and agendas incorporate attention to the needs and circumstances of disadvantaged populations and underrepresented minorities. Mitigating threats to incarcerated people from COVID-19, assuring access to information, prevention, diagnosis, and treatment. Understanding coverage policies (barriers and opportunities) related to testing, treatment, and care.\"\n",
    "\n",
    "list_of_tasks = [task_1, task_2, task_3, task_4, task_5, task_6, task_7, task_8, task_9, task_10]\n",
    "list_of_tasks_short = [task_1_short, task_2_short, task_3_short, task_4_short, task_5_short, task_6_short, task_7_short, task_8_short, task_9_short, task_10_short]\n",
    "\n",
    "#classification with confidence\n",
    "rnd3classes = [2, 0, 3, 0, 3, 7, 7, 7, 6, 5, 4, 5, 0, 0, 0, 0, 4, 5, 5, 1, 0, 1, 1, 1, 1, 7, 7, 3, 3, 3, 2, 2, 3, 3, 9, 2, 2, 3, 3, 2]\n",
    "rnd3confidence = [1, 1, 1, 0.5, 0.5, 1, 1, 0.5, 0.5, 1, 0, 0.5, 1, 0, 1, 1, 0.75, 0.25, 0.25, 0.5, 0.5, 0.75, 0.75, 0.75, 0.75, 1, 0.5, 1, 1, 1, 0, 1, 1, 0, 0, 0.5, 1, 0, 0, 1]\n",
    "\n",
    "\n",
    "stopwords2 = {\n",
    "        'a', 'about', 'above', 'according', 'across', 'after', 'afterwards', 'again', \n",
    "        'against', 'albeit', 'all', 'almost', 'alone', 'along', 'already', 'also', \n",
    "        'although', 'always', 'am', 'among', 'amongst', 'an', 'and', 'another', \n",
    "        'any', 'anybody', 'anyhow', 'anyone', 'anything', 'anyway', 'anywhere', \n",
    "        'apart', 'are', 'around', 'as', 'at', 'av', 'be', 'became', 'because', \n",
    "        'become', 'becomes', 'becoming', 'been', 'before', 'beforehand', 'behind', \n",
    "        'being', 'below', 'beside', 'besides', 'between', 'beyond', 'both', 'but', \n",
    "        'by', 'can', 'cannot', 'canst', 'certain', 'cf', 'choose', 'contrariwise', \n",
    "        'cos', 'could', 'cu', 'day', 'do', 'does', 'doesn', 't', 'doing', 'dost', \n",
    "        'doth', 'double', 'down', 'dual', 'during', 'each', 'either', 'else', \n",
    "        'elsewhere', 'enough', 'et', 'etc', 'even', 'ever', 'every', 'everybody', \n",
    "        'everyone', 'everything', 'everywhere', 'except', 'excepted', 'excepting', \n",
    "        'exception', 'exclude', 'excluding', 'exclusive', 'far', 'farther', \n",
    "        'farthest', 'few', 'ff', 'first', 'for', 'formerly', 'forth', 'forward', \n",
    "        'from', 'front', 'further', 'furthermore', 'furthest', 'get', 'go', 'had', \n",
    "        'halves', 'hardly', 'has', 'hast', 'hath', 'have', 'he', 'hence', \n",
    "        'henceforth', 'her', 'here', 'hereabouts', 'hereafter', 'hereby', 'herein', \n",
    "        'hereto', 'hereupon', 'hers', 'herself', 'him', 'himself', 'hindmost', 'his', \n",
    "        'hither', 'hitherto', 'how','however', 'howsoever', 'i', 'ie', 'if', 'in', \n",
    "        'inasmuch', 'inc', 'include', 'included', 'including', 'indeed', 'indoors', \n",
    "        'inside', 'insomuch', 'instead', 'into', 'inward', 'inwards', 'is', 'it', \n",
    "        'its', 'itself', 'just', 'kg', 'kind', 'km', 'last', 'latter', 'latterly', \n",
    "        'less', 'lest', 'let', 'like', 'little', 'ltd', 'many', 'may', 'maybe', 'me', \n",
    "        'meantime', 'meanwhile', 'might', 'more','moreover', 'most', 'mostly', 'mr', \n",
    "        'mrs', 'ms', 'much', 'must', 'my', 'myself', 'namely', 'need', 'neither', \n",
    "        'never', 'nevertheless', 'next', 'no', 'nobody', 'none', 'nonetheless', \n",
    "        'noone', 'nope', 'nor', 'not', 'nothing', 'notwithstanding', 'now', \n",
    "        'nowadays', 'nowhere', 'of', 'off', 'often', 'ok', 'on', 'once', 'one', \n",
    "        'only', 'onto', 'or', 'other', 'others', 'otherwise', 'ought', 'our', \n",
    "        'ours', 'ourselves', 'out', 'outside', 'over', 'own', 'per', 'perhaps', \n",
    "        'plenty', 'provide', 'quite', 'rather', 'really', 'round', 'said', 'sake', \n",
    "        'same', 'sang', 'save', 'saw', 'see', 'seeing', 'seem', 'seemed', 'seeming', \n",
    "        'seems', 'seen', 'seldom', 'selves', 'sent', 'several', 'shalt', 'she', \n",
    "        'should', 'shown', 'sideways', 'since', 'slept', 'slew', 'slung', 'slunk', \n",
    "        'smote', 'so', 'some', 'somebody', 'somehow', 'someone', 'something', \n",
    "        'sometime', 'sometimes', 'somewhat', 'somewhere', 'spake', 'spat', 'spoke', \n",
    "        'spoken', 'sprang', 'sprung', 'stave', 'staves', 'still', 'such', 'supposing', \n",
    "        'than', 'that', 'the', 'thee', 'their', 'them', 'themselves', 'then', 'thence', \n",
    "        'thenceforth', 'there', 'thereabout', 'thereabouts', 'thereafter', 'thereby', \n",
    "        'therefore', 'therein', 'thereof', 'thereon', 'thereto', 'thereupon', 'these', \n",
    "        'they', 'this', 'those', 'thou', 'though', 'thrice', 'through', 'throughout', \n",
    "        'thru', 'thus', 'thy', 'thyself', 'till', 'to', 'together', 'too', 'toward', \n",
    "        'towards', 'ugh', 'unable', 'under', 'underneath', 'unless', 'unlike', 'until', \n",
    "        'up', 'upon', 'upward', 'upwards', 'us', 'use', 'used', 'using', 'very', 'via', \n",
    "        'vs', 'want', 'was', 'we', 'week', 'well', 'were', 'what', 'whatever', \n",
    "        'whatsoever', 'when', 'whence', 'whenever', 'whensoever', 'where', 'whereabouts', \n",
    "        'whereafter', 'whereas', 'whereat', 'whereby', 'wherefore', 'wherefrom', \n",
    "        'wherein', 'whereinto', 'whereof', 'whereon', 'wheresoever', 'whereto', 'whereunto', \n",
    "        'whereupon', 'wherever', 'wherewith', 'whether', 'whew', 'which', 'whichever', \n",
    "        'whichsoever', 'while', 'whilst', 'whither', 'who', 'whoa', 'whoever', 'whole', \n",
    "        'whom', 'whomever', 'whomsoever', 'whose', 'whosoever', 'why', 'will', 'wilt', \n",
    "        'with', 'within', 'without', 'worse', 'worst', 'would', 'wow', 'ye', 'year', 'yet', \n",
    "        'yippee', 'you', 'your', 'yours', 'yourself', 'yourselves'\n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50\n"
     ]
    }
   ],
   "source": [
    "#read in udel topics, create task_short\n",
    "\n",
    "prefix_t = 'C:/Users/Allemaal/Desktop/ubuntu/Desktop/caos-19/round5/'\n",
    "topics_path = prefix_t + \"topics.covid-round5.xml\"\n",
    "topics_path = prefix_t + \"topics.covid-round5-udel.xml\"\n",
    "write_path = prefix_t + \"topics.covid-round5-ruir.short.xml\"\n",
    "manual_classification = [2, 0, 3, 0, 3, 7, 7, 7, 6, 5, 4, 5, 0, 0, 0, 0, 4, 5, 5, 1, 0, 1, 1, 1, 1, 7, 7, 3, 3, 3, 2, 2, 3, 3, 9, 2, 2, 3, 3, 2, 1, 3, 8, 0, 8, 3, 0, 5, 0, 3]\n",
    "print(len(manual_classification))\n",
    "topics_n = 50\n",
    "\n",
    "tree = Et.parse(topics_path)\n",
    "root = tree.getroot()\n",
    "\n",
    "for i in range(topics_n):\n",
    "    root[i][0].text += \" \" + list_of_tasks_short[manual_classification[i]]\n",
    "\n",
    "#topics = [root[i][0].text + \" \" + list_of_tasks_short[rnd3classes[i]] for i in range(topics_n)]\n",
    "#print(topics)\n",
    "\n",
    "#topics = [t + \" \" + list_of_tasks_short[rnd3classes[i]] for i,t in enumerate(topics)]\n",
    "#print(topics)\n",
    "\n",
    "tree.write(open(write_path, 'wb'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "child index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-58-96312ff019ad>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     10\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     11\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtopics_n\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 12\u001b[1;33m     \u001b[0mroot\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m0\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mtext\u001b[0m \u001b[1;33m+=\u001b[0m \u001b[1;34m\" \"\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mlist_of_tasks_short\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mrnd3classes\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     13\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     14\u001b[0m \u001b[1;31m#topics = [root[i][0].text + \" \" + list_of_tasks_short[rnd3classes[i]] for i in range(topics_n)]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mIndexError\u001b[0m: child index out of range"
     ]
    }
   ],
   "source": [
    "#read in udel topics, create task_only\n",
    "topics_path = prefix_t + \"topics.covid-round3.xml\"\n",
    "write_path = prefix_t + \"topics.covid-round3-ruir.task.short.xml\"\n",
    "topics_n = 50\n",
    "manual_classification = [2, 0, 3, 0, 3, 7, 7, 7, 6, 5, 4, 5, 0, 0, 0, 0, 4, 5, 5, 1, 0, 1, 1, 1, 1, 7, 7, 3, 3, 3, 2, 2, 3, 3, 9, 2, 2, 3, 3, 2, 1, 3, 8, 0, 8, 3, 0, 5, 0, 3]\n",
    "print(len(manual_classification))\n",
    "\n",
    "tree = Et.parse(topics_path)\n",
    "root = tree.getroot()\n",
    "\n",
    "for i in range(topics_n):\n",
    "    root[i][0].text += \" \" + list_of_tasks_short[rnd3classes[i]]\n",
    "\n",
    "#topics = [root[i][0].text + \" \" + list_of_tasks_short[rnd3classes[i]] for i in range(topics_n)]\n",
    "#print(topics)\n",
    "\n",
    "#topics = [t + \" \" + list_of_tasks_short[rnd3classes[i]] for i,t in enumerate(topics)]\n",
    "#print(topics)\n",
    "\n",
    "tree.write(open(write_path, 'wb'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.66\n",
      "[10, 6, 6, 12, 2, 5, 1, 5, 2, 1]\n",
      "[7, 5, 1, 5, 2, 4, 1, 5, 0, 0]\n",
      "transmission incubation environment\n",
      "risk factors\n",
      "genetics origin evolution\n",
      "vaccines therapeutics\n",
      "medical care\n",
      "effectiveness non pharmaceutical interventions\n",
      "diagnostics surveillance\n",
      "geographic spread\n",
      "ethical social considerations\n",
      "information sharing collaboration\n",
      "\n",
      "34\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\"\\nTwo of the tasks have nearly twice as many topics as any other task. The `genetics, origins and evolution' task was \\n\\n- each being twice as popular as any of the alternatives. \\n\\n\\nand the `vaccines and therapeutics' task.\\n\\n\\ntransmission, incubation and environment task. \\n\""
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Let's compare manual and automatic annotation\n",
    "from rank_bm25 import BM25Okapi\n",
    "\n",
    "#create search engine over tasks\n",
    "tokenized_corpus = [doc.split(\" \") for doc in list_of_tasks]\n",
    "bm25 = BM25Okapi(tokenized_corpus)\n",
    "\n",
    "#topics_n = \n",
    "correct = 0\n",
    "selections = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]\n",
    "#iterate over topics\n",
    "for i in range(topics_n):\n",
    "    tokenized_query = root[i][0].text.split(\" \")\n",
    "    doc_scores = bm25.get_scores(tokenized_query)\n",
    "    #print(list(doc_scores).index(max(doc_scores)))\n",
    "    #print(manual_classification[i])\n",
    "    if(list(doc_scores).index(max(doc_scores)) == manual_classification[i]):\n",
    "        correct += 1\n",
    "        \n",
    "    #we also want to know what tasks are popular and if they change over time\n",
    "    selections[manual_classification[i]] += 1\n",
    "\n",
    "print(correct / topics_n)\n",
    "print(selections)\n",
    "\n",
    "#with 30 topics\n",
    "print([7, 5, 1, 5, 2, 4, 1, 5, 0, 0])\n",
    "for item in list_of_tasks_short:\n",
    "    print(item)\n",
    "\n",
    "    \n",
    "    \n",
    "print()\n",
    "for index, val in enumerate(manual_classification):\n",
    "    #print(index)\n",
    "    if(val == 9):\n",
    "        print (index)\n",
    "\n",
    "    \n",
    "    \n",
    "\"\"\"\n",
    "Two of the tasks have nearly twice as many topics as any other task. The `genetics, origins and evolution' task was \n",
    "\n",
    "- each being twice as popular as any of the alternatives. \n",
    "\n",
    "\n",
    "and the `vaccines and therapeutics' task.\n",
    "\n",
    "\n",
    "transmission, incubation and environment task. \n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "OSError",
     "evalue": "[E050] Can't find model 'en_core_sci_sm'. It doesn't seem to be a shortcut link, a Python package or a valid path to a data directory.",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-20-22e5f41b6f93>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[0mtask_ner\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m----> 5\u001b[1;33m \u001b[0mnlp\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mspacy\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mload\u001b[0m\u001b[1;33m(\u001b[0m\u001b[1;34m\"en_core_sci_sm\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m      6\u001b[0m \u001b[1;32mfor\u001b[0m \u001b[0mtask\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mlist_of_tasks\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m      7\u001b[0m     \u001b[0mdoc\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mnlp\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mtask\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\spacy\\__init__.py\u001b[0m in \u001b[0;36mload\u001b[1;34m(name, **overrides)\u001b[0m\n\u001b[0;32m     28\u001b[0m     \u001b[1;32mif\u001b[0m \u001b[0mdepr_path\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32min\u001b[0m \u001b[1;33m(\u001b[0m\u001b[1;32mTrue\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;32mFalse\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;32mNone\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     29\u001b[0m         \u001b[0mwarnings\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mwarn\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mWarnings\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mW001\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpath\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mdepr_path\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mDeprecationWarning\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 30\u001b[1;33m     \u001b[1;32mreturn\u001b[0m \u001b[0mutil\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mload_model\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mname\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0moverrides\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     31\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     32\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m~\\Anaconda3\\lib\\site-packages\\spacy\\util.py\u001b[0m in \u001b[0;36mload_model\u001b[1;34m(name, **overrides)\u001b[0m\n\u001b[0;32m    173\u001b[0m     \u001b[1;32melif\u001b[0m \u001b[0mhasattr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mname\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"exists\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m:\u001b[0m  \u001b[1;31m# Path or Path-like to model data\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    174\u001b[0m         \u001b[1;32mreturn\u001b[0m \u001b[0mload_model_from_path\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mname\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;33m**\u001b[0m\u001b[0moverrides\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m--> 175\u001b[1;33m     \u001b[1;32mraise\u001b[0m \u001b[0mIOError\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mErrors\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mE050\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mname\u001b[0m\u001b[1;33m=\u001b[0m\u001b[0mname\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m    176\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m    177\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mOSError\u001b[0m: [E050] Can't find model 'en_core_sci_sm'. It doesn't seem to be a shortcut link, a Python package or a valid path to a data directory."
     ]
    }
   ],
   "source": [
    "#scispacy NER on tasks to find terms\n",
    "#TODO \n",
    "d\n",
    "task_ner = []\n",
    "nlp = spacy.load(\"en_core_sci_sm\")\n",
    "for task in list_of_tasks:\n",
    "    doc = nlp(task)\n",
    "    ents = \"\"\n",
    "    for e in doc.ents:\n",
    "        ents += \" \" + str(e)\n",
    "    task_ner.append(ents)\n",
    "\n",
    "#create task_long\n",
    "#read in udel topics, create task_short\n",
    "topics_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3.xml\"\n",
    "topics_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3-udel.xml\"\n",
    "write_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3-ruir.long.xml\"\n",
    "topics_n = 40\n",
    "\n",
    "tree = Et.parse(topics_path)\n",
    "root = tree.getroot()\n",
    "\n",
    "for i in range(topics_n):\n",
    "    root[i][0].text += task_ner[rnd3classes[i]]\n",
    "\n",
    "#topics = [root[i][0].text + \" \" + list_of_tasks_short[rnd3classes[i]] for i in range(topics_n)]\n",
    "#print(topics)\n",
    "\n",
    "#topics = [t + \" \" + list_of_tasks_short[rnd3classes[i]] for i,t in enumerate(topics)]\n",
    "#print(topics)\n",
    "\n",
    "tree.write(open(write_path, 'wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load metadata.csv\n",
    "def prepTREC(fname):\n",
    "    #get valid TREC ids for this round\n",
    "    TRECids = []\n",
    "    f = open(fname)\n",
    "    for line in f.readlines():\n",
    "        if line[-1] == '\\n':\n",
    "            line = line[:-1]\n",
    "        TRECids.append(line)\n",
    "    f.close()\n",
    "    \n",
    "    metadata = pd.read_csv(\"/home/tmschoegje/Desktop/caos-19/metadata.csv\")\n",
    "    #now we filter all TREC ids we don't need\n",
    "    #print(np.array(metadata).shape)\n",
    "    #print(np.array(TRECids).shape)\n",
    "    metadata = metadata[metadata.cord_uid.isin(TRECids)]\n",
    "    #there's 33 duplicates (same paper from multiple sources)\n",
    "    metadata.drop_duplicates(subset='cord_uid', keep='first', inplace=True)\n",
    "    \n",
    "    return metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/lib/python3/dist-packages/IPython/core/interactiveshell.py:2822: DtypeWarning: Columns (13,14) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  if self.run_code(code, result):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'transmission': 0.792, 'virus': 0.256, 'health': 0.232}, {'factors': 0.531, 'risk': 0.351, 'transmission': 0.279}, {'evidence': 0.436, 'could': 0.405, 'host': 0.303}, {'studies': 0.408, 'clinical': 0.359, 'development': 0.302}, {'care': 0.599, 'outcomes': 0.408, 'clinical': 0.296}, {'health': 0.493, 'research': 0.305, 'non': 0.288}, {'including': 0.379, 'public': 0.372, 'disease': 0.248}, {'spread': 0.51, 'covid': 0.423, '19': 0.392}, {'health': 0.436, 'measures': 0.433, 'identify': 0.426}, {'public': 0.486, 'risk': 0.414, 'information': 0.407}]\n",
      "[{'transmission': 0.792, 'virus': 0.256, 'health': 0.232}, {'factors': 0.531, 'risk': 0.351, 'transmission': 0.279}, {'evidence': 0.436, 'could': 0.405, 'host': 0.303}, {'studies': 0.408, 'clinical': 0.359, 'development': 0.302}, {'care': 0.599, 'outcomes': 0.408, 'clinical': 0.296}, {'health': 0.493, 'research': 0.305, 'non': 0.288}, {'including': 0.379, 'public': 0.372, 'disease': 0.248}, {'spread': 0.51, 'covid': 0.423, '19': 0.392}, {'health': 0.436, 'measures': 0.433, 'identify': 0.426}, {'public': 0.486, 'risk': 0.414, 'information': 0.407}]\n",
      "[{'transmission': 0.792, 'virus': 0.256, 'health': 0.232}, {'factors': 0.531, 'risk': 0.351, 'transmission': 0.279}, {'evidence': 0.436, 'could': 0.405, 'host': 0.303}, {'studies': 0.408, 'clinical': 0.359, 'development': 0.302}, {'care': 0.599, 'outcomes': 0.408, 'clinical': 0.296}, {'health': 0.493, 'research': 0.305, 'non': 0.288}, {'including': 0.379, 'public': 0.372, 'disease': 0.248}, {'spread': 0.51, 'covid': 0.423, '19': 0.392}, {'health': 0.436, 'measures': 0.433, 'identify': 0.426}, {'public': 0.486, 'risk': 0.414, 'information': 0.407}]\n",
      "[{'transmission': 0.792, 'virus': 0.256, 'health': 0.232}, {'factors': 0.531, 'risk': 0.351, 'transmission': 0.279}, {'evidence': 0.436, 'could': 0.405, 'host': 0.303}, {'studies': 0.408, 'clinical': 0.359, 'development': 0.302}, {'care': 0.599, 'outcomes': 0.408, 'clinical': 0.296}, {'health': 0.493, 'research': 0.305, 'non': 0.288}, {'including': 0.379, 'public': 0.372, 'disease': 0.248}, {'spread': 0.51, 'covid': 0.423, '19': 0.392}, {'health': 0.436, 'measures': 0.433, 'identify': 0.426}, {'public': 0.486, 'risk': 0.414, 'information': 0.407}]\n",
      "[{'transmission': 0.792, 'virus': 0.256, 'health': 0.232}, {'factors': 0.531, 'risk': 0.351, 'transmission': 0.279}, {'evidence': 0.436, 'could': 0.405, 'host': 0.303}, {'studies': 0.408, 'clinical': 0.359, 'development': 0.302}, {'care': 0.599, 'outcomes': 0.408, 'clinical': 0.296}, {'health': 0.493, 'research': 0.305, 'non': 0.288}, {'including': 0.379, 'public': 0.372, 'disease': 0.248}, {'spread': 0.51, 'covid': 0.423, '19': 0.392}, {'health': 0.436, 'measures': 0.433, 'identify': 0.426}, {'public': 0.486, 'risk': 0.414, 'information': 0.407}]\n",
      "[{'transmission': 0.792, 'virus': 0.256, 'health': 0.232}, {'factors': 0.531, 'risk': 0.351, 'transmission': 0.279}, {'evidence': 0.436, 'could': 0.405, 'host': 0.303}, {'studies': 0.408, 'clinical': 0.359, 'development': 0.302}, {'care': 0.599, 'outcomes': 0.408, 'clinical': 0.296}, {'health': 0.493, 'research': 0.305, 'non': 0.288}, {'including': 0.379, 'public': 0.372, 'disease': 0.248}, {'spread': 0.51, 'covid': 0.423, '19': 0.392}, {'health': 0.436, 'measures': 0.433, 'identify': 0.426}, {'public': 0.486, 'risk': 0.414, 'information': 0.407}]\n",
      "[{'transmission': 0.792, 'virus': 0.256, 'health': 0.232}, {'factors': 0.531, 'risk': 0.351, 'transmission': 0.279}, {'evidence': 0.436, 'could': 0.405, 'host': 0.303}, {'studies': 0.408, 'clinical': 0.359, 'development': 0.302}, {'care': 0.599, 'outcomes': 0.408, 'clinical': 0.296}, {'health': 0.493, 'research': 0.305, 'non': 0.288}, {'including': 0.379, 'public': 0.372, 'disease': 0.248}, {'spread': 0.51, 'covid': 0.423, '19': 0.392}, {'health': 0.436, 'measures': 0.433, 'identify': 0.426}, {'public': 0.486, 'risk': 0.414, 'information': 0.407}]\n",
      "[{'transmission': 0.792, 'virus': 0.256, 'health': 0.232}, {'factors': 0.531, 'risk': 0.351, 'transmission': 0.279}, {'evidence': 0.436, 'could': 0.405, 'host': 0.303}, {'studies': 0.408, 'clinical': 0.359, 'development': 0.302}, {'care': 0.599, 'outcomes': 0.408, 'clinical': 0.296}, {'health': 0.493, 'research': 0.305, 'non': 0.288}, {'including': 0.379, 'public': 0.372, 'disease': 0.248}, {'spread': 0.51, 'covid': 0.423, '19': 0.392}, {'health': 0.436, 'measures': 0.433, 'identify': 0.426}, {'public': 0.486, 'risk': 0.414, 'information': 0.407}]\n",
      "[{'transmission': 0.792, 'virus': 0.256, 'health': 0.232}, {'factors': 0.531, 'risk': 0.351, 'transmission': 0.279}, {'evidence': 0.436, 'could': 0.405, 'host': 0.303}, {'studies': 0.408, 'clinical': 0.359, 'development': 0.302}, {'care': 0.599, 'outcomes': 0.408, 'clinical': 0.296}, {'health': 0.493, 'research': 0.305, 'non': 0.288}, {'including': 0.379, 'public': 0.372, 'disease': 0.248}, {'spread': 0.51, 'covid': 0.423, '19': 0.392}, {'health': 0.436, 'measures': 0.433, 'identify': 0.426}, {'public': 0.486, 'risk': 0.414, 'information': 0.407}]\n",
      "[{'transmission': 0.792, 'virus': 0.256, 'health': 0.232}, {'factors': 0.531, 'risk': 0.351, 'transmission': 0.279}, {'evidence': 0.436, 'could': 0.405, 'host': 0.303}, {'studies': 0.408, 'clinical': 0.359, 'development': 0.302}, {'care': 0.599, 'outcomes': 0.408, 'clinical': 0.296}, {'health': 0.493, 'research': 0.305, 'non': 0.288}, {'including': 0.379, 'public': 0.372, 'disease': 0.248}, {'spread': 0.51, 'covid': 0.423, '19': 0.392}, {'health': 0.436, 'measures': 0.433, 'identify': 0.426}, {'public': 0.486, 'risk': 0.414, 'information': 0.407}]\n"
     ]
    }
   ],
   "source": [
    "#let's find the tfidf task terms\n",
    "\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.feature_extraction.text import TfidfTransformer\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.stem import PorterStemmer\n",
    "\n",
    "\n",
    "qrels = []         #contains topic, cord_uid, qrel, assessround qrels\n",
    "doc_sets = []      #contains qrels per task\n",
    "journal_sets = []  #contains individual fulltext abstracts\n",
    "\n",
    "for line in open('/home/tmschoegje/Desktop/caos-19/trecdata/qrels-covid_d3_j0.5-3.txt').readlines():\n",
    "    vals = line.strip().split(\" \")\n",
    "    #if relevant, add\n",
    "    if(float(vals[3]) > 0):\n",
    "        #topic, cord_uid, qrel, assessround\n",
    "        qrels.append([int(vals[0]), vals[2], float(vals[3]), float(vals[1])])\n",
    "qrels = np.array(qrels, dtype=\"O\")\n",
    "\n",
    "sw = stopwords.words(\"english\")\n",
    "porter = PorterStemmer()\n",
    "\n",
    "#finally, we also want to use the other abstracts for our corpus (so our tf idf scores get better)\n",
    "corpus = []\n",
    "metadata = prepTREC('/home/tmschoegje/Desktop/caos-19/trecdata/docids-rnd3.txt')\n",
    "for row in np.array(metadata):\n",
    "    if row[0] not in qrels[:,0]:\n",
    "        corpus.append(str(row[8]))\n",
    "        #corpus += str(row[8])\n",
    "        \n",
    "#Now start computing tf-idf\n",
    "cv = CountVectorizer(max_df=0.85, min_df=0.05, stop_words = sw)\n",
    "word_count_vector = cv.fit_transform(corpus)\n",
    "\n",
    "tfidf_transformer = TfidfTransformer(smooth_idf=True,use_idf=True)\n",
    "tfidf_transformer.fit(word_count_vector)\n",
    "\n",
    "feature_names = cv.get_feature_names()\n",
    "\n",
    "\n",
    "#helper functions for below\n",
    "def sort_coo(coo_matrix):\n",
    "    tuples = zip(coo_matrix.col, coo_matrix.data)\n",
    "    return sorted(tuples, key=lambda x: (x[1], x[0]), reverse=True)\n",
    "    \n",
    "def get_topn(feature_names, sorted_vectors, topn=10):\n",
    "    sorted_vectors = sorted_vectors[:topn]\n",
    "    \n",
    "    scores = []\n",
    "    features = []\n",
    "    \n",
    "    for ind, score in sorted_vectors:\n",
    "        scores.append(round(score, 3))\n",
    "        features.append(feature_names[ind])\n",
    "        \n",
    "    results = {}\n",
    "    for ind in range(len(features)):\n",
    "        results[features[ind]] = scores[ind]\n",
    "        \n",
    "    return results\n",
    "\n",
    "def get_tfidf(task_list, topn):\n",
    "    task_keywords = []\n",
    "    for task_id, task_text in enumerate(task_list):\n",
    "        tfidf_vector = tfidf_transformer.transform(cv.transform([task_text]))\n",
    "        #print(tfidf_vector)\n",
    "\n",
    "        #sort tf idf vectors by scores\n",
    "        sorted_vectors = sort_coo(tfidf_vector.tocoo())\n",
    "\n",
    "        #get top keywords for task w/ score\n",
    "        keywords = get_topn(feature_names, sorted_vectors, topn)\n",
    "        task_keywords.append(keywords)\n",
    "    return task_keywords\n",
    "\n",
    "#    print(keywords)\n",
    "#    task_keywords.append(keywords)\n",
    "\n",
    "for task in list_of_tasks:\n",
    "    print(get_tfidf(list_of_tasks, 3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package stopwords to\n",
      "[nltk_data]     /home/tmschoegje/nltk_data...\n",
      "[nltk_data]   Package stopwords is already up-to-date!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import nltk\n",
    "nltk.download('stopwords')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'transmission': 0.821, 'virus': 0.265, 'health': 0.24, 'disease': 0.202, 'care': 0.185}, {'factors': 0.56, 'risk': 0.37, 'transmission': 0.294, 'virus': 0.285, 'infections': 0.253}, {'evidence': 0.508, 'host': 0.353, 'virus': 0.349, 'infected': 0.319, 'risk': 0.302}, {'clinical': 0.415, 'development': 0.349, 'covid': 0.322, 'studies': 0.314, '19': 0.299}, {'care': 0.558, 'outcomes': 0.435, 'clinical': 0.315, 'management': 0.265, 'data': 0.259}, {'health': 0.552, 'research': 0.342, 'non': 0.323, 'care': 0.319, 'changes': 0.182}, {'public': 0.462, 'disease': 0.308, 'health': 0.275, 'detection': 0.237, 'samples': 0.232}, {'covid': 0.51, '19': 0.474, 'spread': 0.41, 'rate': 0.371, 'virus': 0.264}, {'health': 0.538, 'measures': 0.356, 'public': 0.339, 'outbreak': 0.339, 'research': 0.333}, {'public': 0.496, 'risk': 0.423, 'information': 0.415, 'health': 0.393, 'response': 0.231}]\n",
      "['evidence', 'host', 'virus', 'infected', 'risk']\n",
      "['transmission', 'virus', 'health', 'disease', 'care']\n",
      "['clinical', 'development', 'covid', 'studies', '19']\n",
      "['transmission', 'virus', 'health', 'disease', 'care']\n",
      "['clinical', 'development', 'covid', 'studies', '19']\n",
      "['covid', '19', 'spread', 'rate', 'virus']\n",
      "['covid', '19', 'spread', 'rate', 'virus']\n",
      "['covid', '19', 'spread', 'rate', 'virus']\n",
      "['public', 'disease', 'health', 'detection', 'samples']\n",
      "['health', 'research', 'non', 'care', 'changes']\n",
      "['care', 'outcomes', 'clinical', 'management', 'data']\n",
      "['health', 'research', 'non', 'care', 'changes']\n",
      "['transmission', 'virus', 'health', 'disease', 'care']\n",
      "['transmission', 'virus', 'health', 'disease', 'care']\n",
      "['transmission', 'virus', 'health', 'disease', 'care']\n",
      "['transmission', 'virus', 'health', 'disease', 'care']\n",
      "['care', 'outcomes', 'clinical', 'management', 'data']\n",
      "['health', 'research', 'non', 'care', 'changes']\n",
      "['health', 'research', 'non', 'care', 'changes']\n",
      "['factors', 'risk', 'transmission', 'virus', 'infections']\n",
      "['transmission', 'virus', 'health', 'disease', 'care']\n",
      "['factors', 'risk', 'transmission', 'virus', 'infections']\n",
      "['factors', 'risk', 'transmission', 'virus', 'infections']\n",
      "['factors', 'risk', 'transmission', 'virus', 'infections']\n",
      "['factors', 'risk', 'transmission', 'virus', 'infections']\n",
      "['covid', '19', 'spread', 'rate', 'virus']\n",
      "['covid', '19', 'spread', 'rate', 'virus']\n",
      "['clinical', 'development', 'covid', 'studies', '19']\n",
      "['clinical', 'development', 'covid', 'studies', '19']\n",
      "['clinical', 'development', 'covid', 'studies', '19']\n",
      "['evidence', 'host', 'virus', 'infected', 'risk']\n",
      "['evidence', 'host', 'virus', 'infected', 'risk']\n",
      "['clinical', 'development', 'covid', 'studies', '19']\n",
      "['clinical', 'development', 'covid', 'studies', '19']\n",
      "['public', 'risk', 'information', 'health', 'response']\n",
      "['evidence', 'host', 'virus', 'infected', 'risk']\n",
      "['evidence', 'host', 'virus', 'infected', 'risk']\n",
      "['clinical', 'development', 'covid', 'studies', '19']\n",
      "['clinical', 'development', 'covid', 'studies', '19']\n",
      "['evidence', 'host', 'virus', 'infected', 'risk']\n"
     ]
    }
   ],
   "source": [
    "limit = 5\n",
    "print(get_tfidf(task_ner, limit))\n",
    "tfidf = get_tfidf(task_ner, limit + 3)\n",
    "\n",
    "#read in basic topics, create ruir.qq.tfidf and ruir.qdel.tfidf\n",
    "#topics_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3.xml\"\n",
    "topics_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3-udel.xml\"\n",
    "write_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3-ruir.qdel.tfidf.xml\"\n",
    "topics_n = 40\n",
    "\n",
    "tree = Et.parse(topics_path)\n",
    "root = tree.getroot()\n",
    "\n",
    "for i in range(topics_n):\n",
    "    terms = [term for term in tfidf[rnd3classes[i]]]\n",
    "    \n",
    "    numterms = 0\n",
    "    newterms = []\n",
    "    for t in terms:\n",
    "        if str(t) not in stopwords and numterms < limit:\n",
    "            newterms.append(str(t))\n",
    "            numterms += 1\n",
    "    root[i][0].text += \" \" + \" \".join(newterms)\n",
    "    print(newterms)\n",
    "    \n",
    "#    list_of_tasks_short[rnd3classes[i]]\n",
    "\n",
    "#topics = [root[i][0].text + \" \" + list_of_tasks_short[rnd3classes[i]] for i in range(topics_n)]\n",
    "#print(topics)\n",
    "\n",
    "#topics = [t + \" \" + list_of_tasks_short[rnd3classes[i]] for i,t in enumerate(topics)]\n",
    "#print(topics)\n",
    "\n",
    "tree.write(open(write_path, 'wb'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[' COVID-19 risk factors epidemiological studies Data potential risks factors Smoking pre-existing pulmonary disease Co-infections co-existing respiratory/viral infections virus virulent co-morbidities Neonates pregnant women Socio-economic behavioral factors economic impact virus Transmission dynamics virus reproductive number incubation period serial interval modes transmission environmental factors Severity of disease risk fatality symptomatic hospitalized patients high-risk patient groups Susceptibility populations Public health mitigation effective control', ' COVID-19 risk factors epidemiological studies Data potential risks factors Smoking pre-existing pulmonary disease Co-infections co-existing respiratory/viral infections virus virulent co-morbidities Neonates pregnant women Socio-economic behavioral factors economic impact virus Transmission dynamics virus reproductive number incubation period serial interval modes transmission environmental factors Severity of disease risk fatality symptomatic hospitalized patients high-risk patient groups Susceptibility populations Public health mitigation effective control', ' COVID-19 risk factors epidemiological studies Data potential risks factors Smoking pre-existing pulmonary disease Co-infections co-existing respiratory/viral infections virus virulent co-morbidities Neonates pregnant women Socio-economic behavioral factors economic impact virus Transmission dynamics virus reproductive number incubation period serial interval modes transmission environmental factors Severity of disease risk fatality symptomatic hospitalized patients high-risk patient groups Susceptibility populations Public health mitigation effective control', ' COVID-19 risk factors epidemiological studies Data potential risks factors Smoking pre-existing pulmonary disease Co-infections co-existing respiratory/viral infections virus virulent co-morbidities Neonates pregnant women Socio-economic behavioral factors economic impact virus Transmission dynamics virus reproductive number incubation period serial interval modes transmission environmental factors Severity of disease risk fatality symptomatic hospitalized patients high-risk patient groups Susceptibility populations Public health mitigation effective control', ' COVID-19 risk factors epidemiological studies Data potential risks factors Smoking pre-existing pulmonary disease Co-infections co-existing respiratory/viral infections virus virulent co-morbidities Neonates pregnant women Socio-economic behavioral factors economic impact virus Transmission dynamics virus reproductive number incubation period serial interval modes transmission environmental factors Severity of disease risk fatality symptomatic hospitalized patients high-risk patient groups Susceptibility populations Public health mitigation effective control', ' COVID-19 risk factors epidemiological studies Data potential risks factors Smoking pre-existing pulmonary disease Co-infections co-existing respiratory/viral infections virus virulent co-morbidities Neonates pregnant women Socio-economic behavioral factors economic impact virus Transmission dynamics virus reproductive number incubation period serial interval modes transmission environmental factors Severity of disease risk fatality symptomatic hospitalized patients high-risk patient groups Susceptibility populations Public health mitigation effective control', ' COVID-19 risk factors epidemiological studies Data potential risks factors Smoking pre-existing pulmonary disease Co-infections co-existing respiratory/viral infections virus virulent co-morbidities Neonates pregnant women Socio-economic behavioral factors economic impact virus Transmission dynamics virus reproductive number incubation period serial interval modes transmission environmental factors Severity of disease risk fatality symptomatic hospitalized patients high-risk patient groups Susceptibility populations Public health mitigation effective control', ' COVID-19 risk factors epidemiological studies Data potential risks factors Smoking pre-existing pulmonary disease Co-infections co-existing respiratory/viral infections virus virulent co-morbidities Neonates pregnant women Socio-economic behavioral factors economic impact virus Transmission dynamics virus reproductive number incubation period serial interval modes transmission environmental factors Severity of disease risk fatality symptomatic hospitalized patients high-risk patient groups Susceptibility populations Public health mitigation effective control', ' COVID-19 risk factors epidemiological studies Data potential risks factors Smoking pre-existing pulmonary disease Co-infections co-existing respiratory/viral infections virus virulent co-morbidities Neonates pregnant women Socio-economic behavioral factors economic impact virus Transmission dynamics virus reproductive number incubation period serial interval modes transmission environmental factors Severity of disease risk fatality symptomatic hospitalized patients high-risk patient groups Susceptibility populations Public health mitigation effective control', ' COVID-19 risk factors epidemiological studies Data potential risks factors Smoking pre-existing pulmonary disease Co-infections co-existing respiratory/viral infections virus virulent co-morbidities Neonates pregnant women Socio-economic behavioral factors economic impact virus Transmission dynamics virus reproductive number incubation period serial interval modes transmission environmental factors Severity of disease risk fatality symptomatic hospitalized patients high-risk patient groups Susceptibility populations Public health mitigation effective control']\n"
     ]
    }
   ],
   "source": [
    "print(task_ner)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in udel topics, create task_only\n",
    "topics_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3.xml\"\n",
    "write_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3-ruir.task.short.xml\"\n",
    "topics_n = 40\n",
    "\n",
    "tree = Et.parse(topics_path)\n",
    "root = tree.getroot()\n",
    "\n",
    "for i in range(topics_n):\n",
    "    root[i][0].text += \" \" + list_of_tasks_short[rnd3classes[i]]\n",
    "\n",
    "#topics = [root[i][0].text + \" \" + list_of_tasks_short[rnd3classes[i]] for i in range(topics_n)]\n",
    "#print(topics)\n",
    "\n",
    "#topics = [t + \" \" + list_of_tasks_short[rnd3classes[i]] for i,t in enumerate(topics)]\n",
    "#print(topics)\n",
    "\n",
    "tree.write(open(write_path, 'wb'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\ntmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j3.5-4.txt test.weighted.short.txt \\nmap                   \\tall\\t0.0132\\ntmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j3.5-4.txt test.tfidf3.txt \\nmap                   \\tall\\t0.0145\\ntmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j3.5-4.txt test.tfidf5.txt \\nmap                   \\tall\\t0.0140\\ntmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j0.5-4.txt test.tfidf3.txt \\nmap                   \\tall\\t0.0737\\ntmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j0.5-4.txt test.tfidf5.txt \\nmap                   \\tall\\t0.0672\\ntmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j0.5-4.txt test.weighted.short.txt \\nmap                   \\tall\\t0.0673\\ntmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j0.5-4.txt test.qdel.txt \\nmap                   \\tall\\t0.0580\\n'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read in udel topics, create task_only\n",
    "topics_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3-udel.xml\"\n",
    "write_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3-udel-weighted-short.xml\"\n",
    "topics_n = 40\n",
    "\n",
    "tree = Et.parse(topics_path)\n",
    "root = tree.getroot()\n",
    "\n",
    "for i in range(topics_n):\n",
    "    root[i][0].text += \" \" + root[i][0].text + \" \" + list_of_tasks_short[rnd3classes[i]]\n",
    "\n",
    "#topics = [root[i][0].text + \" \" + list_of_tasks_short[rnd3classes[i]] for i in range(topics_n)]\n",
    "#print(topics)\n",
    "\n",
    "#topics = [t + \" \" + list_of_tasks_short[rnd3classes[i]] for i,t in enumerate(topics)]\n",
    "#print(topics)\n",
    "\n",
    "tree.write(open(write_path, 'wb'))\n",
    "    \n",
    "#TODO vergelijk met/zonder narrative\n",
    "#TODO probeer op nieuwe round data\n",
    "#TODO vergelijk invloed van tasks over de rondes heen\n",
    "#TODO fix anserini influence\n",
    "#manual topic classification based on feedback\n",
    "\n",
    "#This accuracy might decrease over the rounds as the newer , but that in itself could be worth reporting.\n",
    "\n",
    "\n",
    "\"\"\"\n",
    "         0.5-4 map    3.5-4 map     0.5-4 p5    3.5-4 p5    0.5-4 p10   3.5-4 p10   0.5-4 bpref    3.5-4 bpref\n",
    "qq       0.0580       0.0118        0.4550      0.0550      0.3950      0.0550      0.1769         0.1221\n",
    "qq3      0.0521       0.0115        0.4650      0.0700                              0.1650         0.1176\n",
    "qdel     0.0782       0.0144        0.6154      0.0462      0.5564      0.0615      0.2014         0.1439\n",
    "qdel3    0.0737       0.0145        0.5800      0.0550      0.5275      0.0700      0.1980         0.1392\n",
    "qdel5    0.0672       0.0140        0.5400      0.0800      0.4950      0.0700      0.1896         0.1344\n",
    "qshort   0.0673       0.0132        0.4800      0.0650      0.4400      0.0550      0.1950         0.1370\n",
    "\"\"\"\n",
    "\n",
    "\"\"\"\n",
    "tmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j3.5-4.txt test.weighted.short.txt \n",
    "map                   \tall\t0.0132\n",
    "tmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j3.5-4.txt test.tfidf3.txt \n",
    "map                   \tall\t0.0145\n",
    "tmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j3.5-4.txt test.tfidf5.txt \n",
    "map                   \tall\t0.0140\n",
    "tmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j0.5-4.txt test.tfidf3.txt \n",
    "map                   \tall\t0.0737\n",
    "tmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j0.5-4.txt test.tfidf5.txt \n",
    "map                   \tall\t0.0672\n",
    "tmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j0.5-4.txt test.weighted.short.txt \n",
    "map                   \tall\t0.0673\n",
    "tmschoegje@schroefje:~/Desktop/trec_eval/covid$ ./trec_eval -m map qrels-covid_d4_j0.5-4.txt test.qdel.txt \n",
    "map                   \tall\t0.0580\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "50\n",
      "[{'transmission': 0.821, 'virus': 0.265, 'health': 0.24}, {'factors': 0.56, 'risk': 0.37, 'transmission': 0.294}, {'evidence': 0.508, 'host': 0.353, 'virus': 0.349}, {'clinical': 0.415, 'development': 0.349, 'covid': 0.322}, {'care': 0.558, 'outcomes': 0.435, 'clinical': 0.315}, {'health': 0.552, 'research': 0.342, 'non': 0.323}, {'public': 0.462, 'disease': 0.308, 'health': 0.275}, {'covid': 0.51, '19': 0.474, 'spread': 0.41}, {'health': 0.538, 'measures': 0.356, 'public': 0.339}, {'public': 0.496, 'risk': 0.423, 'information': 0.415}]\n",
      "[{'transmission': 0.821, 'virus': 0.265, 'health': 0.24, 'disease': 0.202, 'care': 0.185, 'infection': 0.139, 'blood': 0.103, 'information': 0.102}, {'factors': 0.56, 'risk': 0.37, 'transmission': 0.294, 'virus': 0.285, 'infections': 0.253, 'disease': 0.193, 'impact': 0.151, 'period': 0.151}, {'evidence': 0.508, 'host': 0.353, 'virus': 0.349, 'infected': 0.319, 'risk': 0.302, 'epidemic': 0.187, 'measures': 0.186, 'management': 0.18}, {'clinical': 0.415, 'development': 0.349, 'covid': 0.322, 'studies': 0.314, '19': 0.299, 'viral': 0.289, 'disease': 0.253, 'immune': 0.19}, {'care': 0.558, 'outcomes': 0.435, 'clinical': 0.315, 'management': 0.265, 'data': 0.259, 'medical': 0.177, 'disease': 0.173, 'age': 0.162}, {'health': 0.552, 'research': 0.342, 'non': 0.323, 'care': 0.319, 'changes': 0.182, 'impact': 0.182, 'increase': 0.181, 'spread': 0.178}, {'public': 0.462, 'disease': 0.308, 'health': 0.275, 'detection': 0.237, 'samples': 0.232, 'early': 0.222, 'response': 0.215, 'development': 0.212}, {'covid': 0.51, '19': 0.474, 'spread': 0.41, 'rate': 0.371, 'virus': 0.264, 'mortality': 0.198, 'evidence': 0.192, 'viral': 0.153}, {'health': 0.538, 'measures': 0.356, 'public': 0.339, 'outbreak': 0.339, 'research': 0.333, 'covid': 0.287, '2019': 0.175, 'information': 0.171}, {'public': 0.496, 'risk': 0.423, 'information': 0.415, 'health': 0.393, 'response': 0.231, 'treatment': 0.197, 'data': 0.184, 'measures': 0.174}]\n",
      "hi\n"
     ]
    }
   ],
   "source": [
    "#for reference while manually classifying\n",
    "task_1_short = \"transmission incubation environment\"\n",
    "task_2_short = \"risk factors\"\n",
    "task_3_short = \"genetics origin evolution\"\n",
    "task_4_short = \"vaccines therapeutics\"\n",
    "task_5_short = \"medical care\"\n",
    "task_6_short = \"effectiveness non pharmaceutical interventions\"\n",
    "task_7_short = \"diagnostics surveillance\"\n",
    "task_8_short = \"geographic spread\"\n",
    "task_9_short = \"ethical social considerations\"\n",
    "task_10_short = \"information sharing collaboration\"\n",
    "\n",
    "#manual classification with confidence\n",
    "rnd5classes = [2, 0, 3, 0, 3, 7, 7, 7, 6, 5, 4, 5, 0, 0, 0, 0, 4, 5, 5, 1, 0, 1, 1, 1, 1, 7, 7, 3, 3, 3, 2, 2, 3, 3, 9, 2, 2, 3, 3, 2, 1, 3, 8, 5, 8, 3, 0, 8, 0, 3]\n",
    "rnd5confidence = [1, 1, 1, 0.5, 0.5, 1, 1, 0.5, 0.5, 1, 0, 0.5, 1, 0, 1, 1, 0.75, 0.25, 0.25, 0.5, 0.5, 0.75, 0.75, 0.75, 0.75, 1, 0.5, 1, 1, 1, 0, 1, 1, 0, 0, 0.5, 1, 0, 0, 1]\n",
    "\n",
    "print(len(rnd5classes))\n",
    "\n",
    "\n",
    "mtasks = [\n",
    "    ['environmental transmission', 'incubation', 'contagious', 'persistence', 'stability', 'physical', 'weather',\n",
    "     'epidemiology', 'shedding', 'reproductive number', 'modes of transmission', 'virulent', 'asymptomatic', 'pathogen',\n",
    "     'evolutionary host', 'transmission host'],\n",
    "    ['smoking', 'risk', 'pulmonary', 'pre-condition', 'co-infection', 'high-risk', 'severe', 'susceptible', 'fatality',\n",
    "     'neonates', 'respitory', 'condition', 'pre-existing', 'pregnant', 'morbidities'],\n",
    "    ['human-animal', 'origin', 'genetics', 'evolution', 'genome', 'sample sets', 'genomic', 'strain', 'livestock',\n",
    "     'animal host', 'natural history', 'genetic drift', 'mutation', 'genomics', 'sequencing'],\n",
    "    ['vaccine', 'therapeutic', 'treat', 'drugs', 'pharmaceuticals', 'recipients', 'ADE', 'complication', 'antiviral',\n",
    "     'prophylaxis', 'cloroquine', 'vaccination', 'immume respone'],\n",
    "    ['medical care', 'surge capacity', 'nursing home', 'allocation', 'personal protective equirement',\n",
    "     'clinical characterization', 'nursing', 'care', 'Extracorporeal membrane oxygenation', 'ECMO',\n",
    "     'mechanical ventilation', 'extrapulmonary manifestations', 'cardiomyopathy', 'cardiac arrest',\n",
    "     'regulatory standards', 'N95 masks', 'elastomeric respirators', 'telemedicine', 'steroids', 'high flow oxygen',\n",
    "     'supportive interventions'],\n",
    "    ['NPI', 'non-pharmaceutical intervention', 'school closure', 'travel ban', 'quarantine', 'mass gathering',\n",
    "     'social distancing', 'public health advice', 'economic impact'],\n",
    "    ['counties', 'geographic', 'geography', 'mortality rate', 'spread', 'mutations'],\n",
    "    ['diagnostic', 'surveillance', 'detection', 'screening', 'ELISAs', 'capacity', 'testing', 'point-of-care',\n",
    "     'rapid testing', 'pathogen', 'reagent', 'cytokines', 'response markers', 'swabs'],\n",
    "    ['ethical', 'social science', 'principles', 'standards', 'ethics', 'psychological health', 'fear', 'anxiety',\n",
    "     'stigma', 'sociology'],\n",
    "    ['collaboration', 'nomenclature', 'data standards', 'information sharing', 'communication', 'collaborate',\n",
    "     'coordination', 'misunderstanding', 'action plan']\n",
    "]\n",
    "\n",
    "\n",
    "limit = 3\n",
    "print(get_tfidf(task_ner, limit))\n",
    "tfidf = get_tfidf(task_ner, limit + 5)\n",
    "\n",
    "#VERSIONS  note these are all extending qq\n",
    "#terms: tfidf3, tfidf5, mtasks, list_of_tasks_short\n",
    "#weights: 1x, 2x, 3x\n",
    "#e.g. ruir31, ruir51, ruirm2, ruirs3\n",
    "\n",
    "\n",
    "#read in basic topics, create ruir.qq.tfidf and ruir.qdel.tfidf\n",
    "topics_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round5.xml\"\n",
    "topics_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round5-udel.xml\"\n",
    "#write_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round5-ruir.qdel.weight2.tfidf\" + str(limit) + \".xml\"\n",
    "write_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round5-udel33.xml\"\n",
    "topics_n = 50\n",
    "\n",
    "tree = Et.parse(topics_path)\n",
    "root = tree.getroot()\n",
    "print(tfidf)\n",
    "print('hi')\n",
    "\n",
    "for i in range(topics_n):\n",
    "    terms = [term for term in list_of_tasks_short[rnd5classes[i]].split(\" \")]#mtasks[rnd5classes[i]]]#tfidf[rnd5classes[i]]]\n",
    "    #print(terms)\n",
    "    \n",
    "    numterms = 0\n",
    "    newterms = []\n",
    "    for t in terms:\n",
    "        if str(t) not in stopwords2 and numterms < limit:\n",
    "            newterms.append(str(t))\n",
    "            numterms += 1\n",
    "    root[i][0].text += \" \" + root[i][0].text + \" \" + root[i][0].text + \" \" + \" \".join(newterms)\n",
    "    #print(newterms)\n",
    "    \n",
    "#    list_of_tasks_short[rnd3classes[i]]\n",
    "\n",
    "#topics = [root[i][0].text + \" \" + list_of_tasks_short[rnd3classes[i]] for i in range(topics_n)]\n",
    "#print(topics)\n",
    "\n",
    "#topics = [t + \" \" + list_of_tasks_short[rnd3classes[i]] for i,t in enumerate(topics)]\n",
    "#print(topics)\n",
    "\n",
    "tree.write(open(write_path, 'wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'transmission': 0.821, 'virus': 0.265, 'health': 0.24}, {'factors': 0.56, 'risk': 0.37, 'transmission': 0.294}, {'evidence': 0.508, 'host': 0.353, 'virus': 0.349}, {'clinical': 0.415, 'development': 0.349, 'covid': 0.322}, {'care': 0.558, 'outcomes': 0.435, 'clinical': 0.315}, {'health': 0.552, 'research': 0.342, 'non': 0.323}, {'public': 0.462, 'disease': 0.308, 'health': 0.275}, {'covid': 0.51, '19': 0.474, 'spread': 0.41}, {'health': 0.538, 'measures': 0.356, 'public': 0.339}, {'public': 0.496, 'risk': 0.423, 'information': 0.415}]\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "argument of type 'WordListCorpusReader' is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-12-b35f5aa6b128>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     19\u001b[0m     \u001b[0mnewterms\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     20\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mt\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mterms\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 21\u001b[0;31m         \u001b[0;32mif\u001b[0m \u001b[0mt\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mstopwords\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mnumterms\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mlimit\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     22\u001b[0m             \u001b[0mnewterms\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mt\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m             \u001b[0mnumterms\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: argument of type 'WordListCorpusReader' is not iterable"
     ]
    }
   ],
   "source": [
    "#now qq + tfidf\n",
    "limit = 3\n",
    "print(get_tfidf(task_ner, limit))\n",
    "tfidf = get_tfidf(task_ner, limit + 5)\n",
    "\n",
    "#read in basic topics, create ruir.qq.tfidf and ruir.qdel.tfidf\n",
    "#topics_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3.xml\"\n",
    "topics_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3-udel.xml\"\n",
    "write_path = \"/home/tmschoegje/Desktop/caos-19/round5/topics.covid-round3-ruir.qq.weighted.tfidf\" + str(limit) + \".xml\"\n",
    "topics_n = 40\n",
    "\n",
    "tree = Et.parse(topics_path)\n",
    "root = tree.getroot()\n",
    "\n",
    "for i in range(topics_n):\n",
    "    terms = [term for term in tfidf[rnd3classes[i]]]\n",
    "    \n",
    "    numterms = 0\n",
    "    newterms = []\n",
    "    for t in terms:\n",
    "        if t not in stopwords and numterms < limit:\n",
    "            newterms.append(t)\n",
    "            numterms += 1\n",
    "    root[i][0].text += \" \" + root[i][0].text + \" \" + root[i][0].text + \" \" + \" \".join(newterms)\n",
    "    print(newterms)\n",
    "    \n",
    "#    list_of_tasks_short[rnd3classes[i]]\n",
    "\n",
    "#topics = [root[i][0].text + \" \" + list_of_tasks_short[rnd3classes[i]] for i in range(topics_n)]\n",
    "#print(topics)\n",
    "\n",
    "#topics = [t + \" \" + list_of_tasks_short[rnd3classes[i]] for i,t in enumerate(topics)]\n",
    "#print(topics)\n",
    "\n",
    "tree.write(open(write_path, 'wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#test\n",
    "#quick query term test\n",
    "from pyserini.search import SimpleSearcher\n",
    "#from pyserini.index import pyutils\n",
    "#from pyserini.analysis import pyanalysis\n",
    "#from pyserini.search import pysearch\n",
    "#from pyserini.search import pyquerybuilder\n",
    "#from pyserini.pyclass import autoclass\n",
    "#from pyserini.analysis.pyanalysis import get_lucene_analyzer\n",
    "\n",
    "index_loc = '/home/tmschoegje/Desktop/anserini/indexes/lucene-index-cord19-full-text-2020-05-19/'\n",
    "searcher = SimpleSearcher(index_loc)\n",
    "#index_utils = pyutils.IndexReaderUtils(index_loc)\n",
    "\n",
    "\n",
    "hits = searcher.search('corona origin')\n",
    "\n",
    "# Print the first 10 hits:\n",
    "for i in range(0, 10):\n",
    "    print(f'{i+1:2} {hits[i].docid:15} {hits[i].score:.5f}')\n",
    "    \n",
    "    \n",
    "hits = searcher.search('corona origin origin origin')\n",
    "print()\n",
    "for i in range(0, 10):\n",
    "    print(f'{i+1:2} {hits[i].docid:15} {hits[i].score:.5f}')\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
